src/resampling_crop_type.py

```diff
@@ -19,71 +24,129 @@ def __init__(self, ROOT_DIR:str) -> None:
-        """
-        # use the new dataset to define the desired sample size
-        self.n_class_1 = self.df_label_comb[self.df_label_comb.label==1].shape[0]
-        self.class_1 = self.df_label_comb[self.df_label_comb.label==1]
-        # from the new dataset  create the subset of classes i want to oversample
-        df_over = self.df_label_comb.loc[self.df_label_comb['label'].isin([5,6,3,8])]
-        #from our subset define the target and features
-        X_over= df_over.drop('label',axis=1)
-        y_over = df_over['label']
-        # oversample the X_under and y_under
-        strategy = { 8:self.n_class_1, 3:self.n_class_1, 5:self.n_class_1, 6:self.n_class_1}
-        over = RandomOverSampler(sampling_strategy=strategy)
-        X_overS, y_overS = over.fit_resample(X_over, y_over)
-        print(f"overSampling: {Counter(y_overS)}")
-        # create a set of oversampled data by merging the above sampled feature and Target
-        self.df_overS = pd.concat([X_overS,y_overS],axis= 1)
+    def get_base_label_df(self, df_label_comb:pd.DataFrame, label_ranking:pd.DataFrame
+        ) -> pd.DataFrame:
+        # get the data for the label to resample on (base label)
+        sampling_base_label_idx = self.get_base_label_index(label_ranking)
+        sampling_base_label = label_ranking.iloc[sampling_base_label_idx]["label"]
+        df_base_label = df_label_comb[df_label_comb["label"]==sampling_base_label]
+        return df_base_label
 
-    def undersampling(self):
-        """ Undersamples the overrepresented crop classes.
-        """
-        # repeat the same thing for undersampling
-        df_under = self.df_label_comb.loc[self.df_label_comb['label'].isin([4,2,7])]
-        X_under= df_under.drop('label',axis=1)
-        y_under = df_under['label']
-        strategy = {4:self.n_class_1,2:self.n_class_1,7:self.n_class_1}
-        under = RandomUnderSampler(sampling_strategy=strategy)
-        X_underS, y_underS = under.fit_resample(X_under, y_under)
-        print(f"underSampling: {Counter(y_underS)}")
-        self.df_underS = pd.concat([X_underS,y_underS],axis= 1)
+    def get_resampling_size(self, label_ranking:pd.DataFrame
+        ) -> int:
+        # get the index of the label to resample on and set the sample size to that label
+        sampling_base_label_idx = self.get_base_label_index(label_ranking) 
+        sampling_size = int(label_ranking.iloc[sampling_base_label_idx]["field_id"])
+        return sampling_size 
 
-    def save_train_test_data(self):
+    def get_labels_to_resample(self, label_ranking:pd.DataFrame
+        ) -> Union[pd.Series, pd.Series]:
+        # get the classes for which we need to do over- and undersampling
+        sampling_base_label_idx = self.get_base_label_index(label_ranking) 
+        classes_to_oversample = label_ranking[0:sampling_base_label_idx]["label"]
+        classes_to_undersample = label_ranking[sampling_base_label_idx+1:len(label_ranking)]["label"]
```

Comment:
transform this label_ranking[sampling_base_label_idx+1:len(label_ranking)]["label"]

to 

 label_ranking[sampling_base_label_idx+1:-1]["label"]